### PR TITLE
Enhancement/Refactoring to reduce class coupling and improve overall structure

### PR DESCRIPTION
**Pull request by cmatgh**
- Add interfaces for every operations class
- Add abstract factory for the operations and implement a factory for the default operations
- Reduce class coupling by injecting factory into constructor calls for logic classes
- Add Logging Decorators for all operation classes

**Note**:
In agreement with Felix we have decided to put this PR on hold, since it involves a huge refactoring that is not necessary for the final release but could be added at a later time to improve the overall structure.

### CODE REVIEW

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/RenameLogic.java [hunk 1]
```diff
@@ -3,27 +3,45 @@
...
+    }
+
     /**
      * Rename an existing application.
-     * @param applicationsOperations ApplicationsOperations
+
      * @param newName The new name of the application
      * @param currentName The current name of the application
      * @throws UpdateException if an error occurs during the nameChange procedure
      * @throws NullPointerException when newName, currentName or applicationsOperations is null
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/RenameLogic.java:41]: The applicationsOperations are no longer passed:
```suggestion
     * @throws NullPointerException if any of the arguments is null
```

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/RenameLogic.java [hunk 2]
```diff
@@ -36,14 +54,13 @@ public void renameApplication(ApplicationsOperations applicationsOperations, Str
 
     /**
      * Rename an existing service.
-     * @param servicesOperations ServicesOperations
      * @param newName The new name of the service
      * @param currentName The current name of the service
      * @throws UpdateException if an error occurs during the nameChange procedure
      * @throws NullPointerException when newName, currentName or servicesOperations is null
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/logic/RenameLogic.java:60]: The servicesOperations are no longer passed:
```suggestion
     * @throws NullPointerException if any of the arguments is null
```

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ApplicationsOperations.java [hunk 3]
```diff
@@ -1,545 +1,40 @@
...
 
 /**
- * Handles the operations for querying and manipulating applications on a cloud
- * foundry instance.
- *
- * To retrieve the data from resulting Mono or Flux objects you can use
- * subscription methods (block, subscribe, etc.) provided by the reactor
- * library. For more details on how to work with Mono's visit:
- * https://projectreactor.io/docs/core/release/reference/index.html#core-features
+ * Serves has the pattern for the application operations classes
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ApplicationsOperations.java:10]: The sentence is a bit broken

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ApplicationsOperations.java [hunk 4]
```diff
@@ -1,545 +1,40 @@
...
 
-    private DomainSummary createDomainSummary(Domain domain) {
-        return DomainSummary
-                .builder()
-                .id(domain.getId())
-                .name(domain.getName())
-                .type(domain.getType())
-                .build();
-    }
+    Mono<Void> removeRoute(String applicationName, String route);
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ApplicationsOperations.java:38]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ClientOperations.java [hunk 5]
```diff
@@ -1,35 +1,12 @@
...
-import org.cloudfoundry.operations.DefaultCloudFoundryOperations;
 import reactor.core.publisher.Mono;
 
 /**
- * Handles the operations to determine meta-information from a cloud foundry instance.
- * <p>
- * To retrieve the data from resulting Mono or Flux objects you can use subscription methods (block, subscribe, etc.)
- * provided by the reactor library. For more details on how to work with Mono's visit:
- * https://projectreactor.io/docs/core/release/reference/index.html#core-features
+ * Serves has the pattern for the client operations classes
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ClientOperations.java:6]: The sentence is a bit broken

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ClientOperations.java [hunk 6]
```diff
@@ -1,35 +1,12 @@
...
-     *
-     * @return API-Version
-     */
-    public Mono<String> determineApiVersion() {
-        CloudFoundryClient cfClient = cloudFoundryOperations.getCloudFoundryClient();
-        GetInfoRequest infoRequest = GetInfoRequest.builder().build();
-        Info cfClientInfo = cfClient.info();
-        return cfClientInfo.get(infoRequest).map(GetInfoResponse::getApiVersion);
-    }
+    Mono<String> determineApiVersion();
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ClientOperations.java:10]: A brief documentation should be added to the method

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/OperationsFactory.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    public abstract ApplicationsOperations createApplicationsOperations();
+
+    public abstract ServicesOperations createServiceOperations();
+
+    public abstract SpaceDevelopersOperations createSpaceDevelopersOperations();
+
+    public abstract SpaceOperations createSpaceOperations();
+
+    public abstract ClientOperations createClientOperations();
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/OperationsFactory.java:26]: A brief documentation should be added to the methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ServicesOperations.java [hunk 8]
```diff
@@ -1,335 +1,30 @@
...
-import java.util.Collections;
 import java.util.Map;
 
 /**
- * Handles the operations for querying and manipulating services on a cloud foundry instance.
- *
- * To retrieve the data from resulting Mono or Flux objects you can use subscription methods (block, subscribe, etc.)
- * provided by the reactor library. For more details on how to work with Mono's visit:
- * https://projectreactor.io/docs/core/release/reference/index.html#core-features
+ * Serves has the pattern for the service operations classes
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ServicesOperations.java:10]: The sentence is a bit broken

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ServicesOperations.java [hunk 9]
```diff
@@ -1,335 +1,30 @@
...
 
-    private UnbindRouteServiceInstanceRequest createRouteServiceInstanceRequest(Route route) {
-        return UnbindRouteServiceInstanceRequest
-                .builder()
-                .serviceInstanceName(route.getService())
-                .domainName(route.getDomain())
-                .hostname(route.getHost())
-                .build();
-    }
+    Flux<Void> unbindRoutes(String serviceInstanceName);
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ServicesOperations.java:28]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceDevelopersOperations.java [hunk 10]
```diff
@@ -1,116 +1,21 @@
...
 import java.util.List;
 
-
 /**
- * Handles the operations for manipulating space developers on a cloud foundry instance.
- *
- * To retrieve the data from resulting Mono or Flux objects you can use subscription methods (block, subscribe, etc.)
- * provided by the reactor library. For more details on how to work with Mono's visit:
- * https://projectreactor.io/docs/core/release/reference/index.html#core-features
+ * Serves has the pattern for the space developers operations classes
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceDevelopersOperations.java:9]: The sentence is a bit broken

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceDevelopersOperations.java [hunk 11]
```diff
@@ -1,116 +1,21 @@
...
 
-        return cloudFoundryOperations.getCloudFoundryClient()
-                .spaces()
-                .removeDeveloperByUsername(request)
-                .doOnSubscribe(subscription -> log.debug("Removing space developer", username))
-                .doOnSuccess(subscription -> log.verbose("Removing space developer", username, "completed"))
-                .onErrorStop()
-                .then(Mono.empty());
-    }
+    Mono<Void> remove(String username, String spaceId);
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceDevelopersOperations.java:19]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceOperations.java [hunk 12]
```diff
@@ -1,56 +1,16 @@
...
-        return this.cloudFoundryOperations
-                .spaces()
-                .list()
-                // ignore space ids for now (they could be cached to avoid querying them again,
-                // when manipulating space developers)
-                .map(SpaceSummary::getName)
-                .collectList();
-    }
+/**
+ * Serves has the pattern for the space operations classes
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceOperations.java:8]: The sentence is a bit broken

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceOperations.java [hunk 13]
```diff
@@ -1,56 +1,16 @@
...
+    Mono<List<String>> getAll();
 
-        CreateSpaceRequest createSpaceRequest = CreateSpaceRequest.builder()
-                .name(spaceName)
-                .build();
-        return this.cloudFoundryOperations.spaces().create(createSpaceRequest)
-                .doOnSubscribe(aVoid -> log.debug("Creating space", spaceName))
-                .doOnSuccess(aVoid -> log.verbose("Creating space", spaceName, "completed"));
-    }
+    Mono<Void> create(String spaceName);
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/SpaceOperations.java:14]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+import java.util.Map;
+
+/**
+ * This class utilizes the decorator pattern to apply logging to the underlying applications operations
+ */
+public class ApplicationsOperationsLogging implements ApplicationsOperations {
+
+    private final Log log;
+
+    ApplicationsOperations applicationsOperations;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java:21]: Can this field be set to `private`/`protected`?

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+                        "Adding route", route, "to app", applicationName, "completed"));
+    }
+
+    @Override
+    public Mono<Void> removeRoute(String applicationName, String route) {
+        return this.applicationsOperations.removeRoute(applicationName, route)
+                .doOnSubscribe(aVoid -> log.info("Removing route", route, "from app", applicationName))
+                .doOnSuccess(aVoid -> log.info(
+                        "Removing route", route, "from app", applicationName, "completed"));
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java:150]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+import cloud.foundry.cli.crosscutting.logging.Log;
+import cloud.foundry.cli.crosscutting.mapping.beans.ServiceBean;
+import cloud.foundry.cli.operations.ServicesOperations;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+public class ServicesOperationsLogging implements ServicesOperations {
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java:12]: A documentation explaining the purpose of this class would be nice

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+                        "Unbinding all applications from service instance", serviceInstanceName, "completed"));
+    }
+
+    @Override
+    public Flux<Void> unbindRoutes(String serviceInstanceName) {
+        return this.servicesOperations.unbindRoutes(serviceInstanceName)
+                .doOnSubscribe(aVoid -> log.info("Unbinding all routes from service instance", serviceInstanceName))
+                .doOnComplete(() -> log.verbose(
+                        "Unbinding all routes from service instance", serviceInstanceName, "completed"));
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java:89]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/DefaultSpaceOperations.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+import cloud.foundry.cli.operations.AbstractOperations;
+import cloud.foundry.cli.operations.SpaceOperations;
+import org.cloudfoundry.operations.DefaultCloudFoundryOperations;
+import org.cloudfoundry.operations.spaces.CreateSpaceRequest;
+import org.cloudfoundry.operations.spaces.SpaceSummary;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+
+public class DefaultSpaceOperations extends AbstractOperations<DefaultCloudFoundryOperations>
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/DefaultSpaceOperations.java:15]: A documentation explaining the purpose of this class would be nice

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+import cloud.foundry.cli.operations.ServicesOperations;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+public class ServicesOperationsLogging implements ServicesOperations {
+
+    private final Log log;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/services/ServicesOperationsLogging.java:14]: Usually the log instance is `static` and initialized right away

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+/**
+ * This class utilizes the decorator pattern to apply logging to the underlying applications operations
+ */
+public class ApplicationsOperationsLogging implements ApplicationsOperations {
+
+    private final Log log;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/applications/ApplicationsOperationsLogging.java:19]: Usually the log instance is `static` and initialized right away

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+import cloud.foundry.cli.crosscutting.logging.Log;
+import cloud.foundry.cli.operations.SpaceOperations;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public class SpaceOperationsLogging implements SpaceOperations {
+
+    private final Log log;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java:12]: Usually the log instance is `static` and initialized right away

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java [hunk 22]
```diff
@@ -0,0 +1,34 @@
+package cloud.foundry.cli.operations.space;
+
+import cloud.foundry.cli.crosscutting.logging.Log;
+import cloud.foundry.cli.operations.SpaceOperations;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public class SpaceOperationsLogging implements SpaceOperations {
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java:10]: A documentation explaining the purpose of this class would be nice

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+
+    private final Log log;
+
+    public SpaceOperations spaceOperations;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java:14]: Does this field really have to be public?

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+                .doOnSubscribe(subscription -> log.debug("Querying all spaces"))
+                .doOnSuccess(stringApplicationBeanMap -> log.debug("Querying all spaces finished"));
+    }
+
+    @Override
+    public Mono<Void> create(String spaceName) {
+        return this.spaceOperations.create(spaceName)
+                .doOnSubscribe(aVoid -> log.debug("Creating space", spaceName))
+                .doOnSuccess(aVoid -> log.verbose("Creating space", spaceName, "completed"));
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/space/SpaceOperationsLogging.java:33]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+import cloud.foundry.cli.crosscutting.logging.Log;
+import cloud.foundry.cli.operations.SpaceDevelopersOperations;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public class SpaceDevelopersOperationsLogging implements SpaceDevelopersOperations {
+
+    private final Log log;
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java:12]: Usually the log instance is static and initialized right away

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java [hunk 26]
```diff
@@ -0,0 +1,49 @@
+package cloud.foundry.cli.operations.spacedevelopers;
+
+import cloud.foundry.cli.crosscutting.logging.Log;
+import cloud.foundry.cli.operations.SpaceDevelopersOperations;
+import reactor.core.publisher.Mono;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+public class SpaceDevelopersOperationsLogging implements SpaceDevelopersOperations {
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java:10]: A documentation explaining the purpose of this class would be nice

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java [hunk 27]
```diff
@@ -0,0 +1,49 @@
...
+                .doOnSubscribe(subscription -> log.debug("Assigning space developer", username))
+                .doOnSuccess(subscription -> log.verbose("Assigning space developer", username, "completed"));
+    }
+
+    @Override
+    public Mono<Void> remove(String username, String spaceId) {
+        return this.spaceDevelopersOperations.remove(username, spaceId)
+                .doOnSubscribe(subscription -> log.debug("Removing space developer", username))
+                .doOnSuccess(subscription -> log.verbose("Removing space developer", username, "completed"));
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/spacedevelopers/SpaceDevelopersOperationsLogging.java:48]: A brief documentation should be added to these methods

File: cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/OperationsFactory.java [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+
+    private static OperationsFactory instance;
+
+    public static OperationsFactory getInstance() {
+        return instance;
+    }
+
+    public static void setInstance(OperationsFactory instance) {
+        OperationsFactory.instance = instance;
+    }
```
Comment (felix-oq) [inline cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/OperationsFactory.java:16]: Why is this needed?

**Comment by cmatgh** [comment]
In agreement with Felix, we have decided to put this PR on hold. Since it involves a huge refactoring that is not necessary for the final release but could be added at a later time to improve the overall structure.
